server/server.go

```diff
@@ -2,41 +2,68 @@ package server
 
 import (
 	"net/http"
-	"github.com/gruntwork-io/health-checker/options"
 	"net"
 	"fmt"
+	"sync"
+	"time"
+	"github.com/gruntwork-io/health-checker/options"
+	"github.com/gruntwork-io/gruntwork-cli/errors"
 )
 
-func StartHttpServer(opts *options.Options) {
+type httpResponse struct {
+	StatusCode int
+	Body string
+}
+
+func StartHttpServer(opts *options.Options) error {
 	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
-		checkTcpPorts(w, r, opts)
+		resp := checkTcpPorts(opts)
+		err := writeHttpResponse(w, resp)
+		if err != nil {
+			opts.Logger.Error("Failed to send HTTP response. Exiting.")
+			panic(err)
+		}
 	})
-	http.ListenAndServe(opts.Listener, nil)
+	err := http.ListenAndServe(opts.Listener, nil)
+	if err != nil {
+		return err
+	}
+
+	return nil
 }
 
 // Check that we can open a TPC connection to all the ports in opts.Ports
-func checkTcpPorts(w http.ResponseWriter, r *http.Request, opts *options.Options) {
+func checkTcpPorts(opts *options.Options) *httpResponse {
 	logger := opts.Logger
 	logger.Infof("Received inbound request. Beginning health checks...")
 
 	allPortsValid := true
 
+	var waitGroup = sync.WaitGroup{}
+
 	for _, port := range opts.Ports {
-		err := attemptTcpConnection(port, opts)
-		if err != nil {
-			logger.Warnf("TCP connection to port %d FAILED!", port)
-			allPortsValid = false
-		} else {
-			logger.Infof("TCP connection to port %d successful", port)
-		}
+		waitGroup.Add(1)
+		go func(port int) {
+			err := attemptTcpConnection(port, opts)
+			if err != nil {
+				logger.Warnf("TCP connection to port %d FAILED: %s", port, err)
+				allPortsValid = false
```

Comment:
This is not safe to do in a concurrent environment! You're modifying shared state from multiple different goroutines at the same time.

Use an integer (initialized to 0) instead with an [atomic increment](https://golang.org/pkg/sync/atomic/) operation on errors. If it ends up greater than 0, at least one health check failed.

Reply:
Also, there should be tests for all of this.